### PR TITLE
Piecewise recursively evaluates itself

### LINKED ISSUE #17165: Piecewise simplification (to avoid doit growth)
```python
>>> Piecewise((1,x<1),(2,True))
Piecewise((1, x < 1), (2, True))
>>> p1=_
>>> p2 = Piecewise((1,x<1),(p1,True))
Piecewise((1, x < 1), (Piecewise((1, x < 1), (2, True)), True))
>>> piecewise_fold(_)  <-- should this have been necessary?
Piecewise((1, x < 1), (2, True))
```
Perhaps that situation should be recognized during instantiation. Or else `doit` might need to call `piecewise_fold` first to avoid repeating what is known (as was reported [on the mailing list](https://groups.google.com/forum/#!topic/sympy/-39qnrULloM)):
```python
>>> Sum (x**n, (n, -1, oo)).doit () 
Piecewise((1/(x*(1 - x)), Abs(x) < 1), (Sum(x**n, (n, -1, oo)), True))
>>> _.doit()
Piecewise((1/(x*(1 - x)), Abs(x) < 1), (Piecewise((1/(x*(1 - x)), Abs(x) < 1), (Sum(x**n, (n, -1, oo)), True)), True))
```

### PR DESCRIPTION
**Pull request by tom-pytel**
#### References to other Issues or PRs
Fixes #17165

#### Brief description of what is fixed or changed
Changed Piecewise.doit() to avoid recursively evaluating itself ad infinitium with each execution.

#### Other comments
This only partially addresses #17165, the other issue being whether the constructor should fold a piecewise expression upon instantiation to avoid a recursive nesting. In that case maybe it is best to leave it to the user to decide whether they want to fold or no? Or add a hint to request folding from the constructor?

#### Release Notes

* functions
  * Changed Piecewise.doit() to avoid recursively evaluating itself.

### CODE REVIEW

File: sympy/matrices/expressions/matpow.py [hunk 1]
```diff
@@ -71,7 +71,7 @@ def doit(self, **kwargs):
             raise ValueError("Matrix determinant is 0, not invertible.")
         elif isinstance(base, (Identity, ZeroMatrix)):
             return base
-        elif isinstance(base, MatrixBase) and exp.is_number:
+        elif isinstance(base, MatrixBase) and (exp.is_number or exp.is_negative is not None or base.det() != 0):
```
Comment (oscarbenjamin) [inline sympy/matrices/expressions/matpow.py]: Is this change a crossover from the other PR?

**Comment by Pristine-Cat** [comment]
Yes it appears to be, my bad...

File: sympy/matrices/expressions/matpow.py (hunk 1 above)
Comment (smichr) [reply sympy/matrices/expressions/matpow.py]: #17175, perhaps

**Comment by smichr** [comment]
I wonder if this situation should be raising an error instead. A doit shouldn't return a self-referent expression. In the case of the issue's example, Sum when abs(x)<1 is defined and when abs(x)>=1 then is it not defined? Then Undefined should be given as the 2nd argument. By raising an error in Piecewise, it would catch such errors in rewriting functions.

**Comment by Pristine-Cat** [comment]
> I wonder if this situation should be raising an error instead. A doit 
> shouldn't return a self-referent expression. In the case of the 
> issue's example, Sum when abs(x)<1 is defined and when abs(x)>=1 then 
> is it not defined? Then Undefined should be given as the 2nd argument. 
> By raising an error in Piecewise, it would catch such errors in 
> rewriting functions.
>
It doesn't acually return a self-referent expression but rather one with 
an identical copy of the parent, which upon the next evaluation will get 
another identical copy of the parent nested within that - think 
Matryoshka dolls. Having something noted as Undefined under some 
conditions but defined for other seems exactly like what Piecewise is 
supposed to do so I don't think it should be raising for Undefined, but 
simply not try to evaluate it into another Piecewise if there is no 
information that this would add to the expression (it is already present 
in the parent).

**Comment by oscarbenjamin** [comment]
Ideally Piecewise would be able to refine subexpressions based on the conditions they are required to satisfy which would help with this. In this particular case though piecewise_fold leads to cases with identical conditions which are automatically filtered Piecewise.eval
```
In [23]: Piecewise((x, x>1), (0, x>1), (1, True))                                                                                 
Out[23]: 
⎧x  for x > 1
⎨            
⎩1  otherwise
```
Is there any reason not to piecewise_fold in Piecewise.eval or Piecewise.doit?

**Comment by smichr** [comment]
And advantage to having the unevaluated object that created the Piecewise *in* the Piecewise is that you can see where the evaluated form came from: `Piecewise((possible form, under these conditions), (initial form, otherwise))`. I added a test. 

I think this looks ok.

**Comment by smichr** [comment]
> I think this looks ok.

The reason I think this step is correct is because folding is different than doing...it's a simplification step whereas doit is just a form-changing step.

**Comment by Pristine-Cat** [comment]
I like it because it is simple and efficient, though I could see folding 
being used at initialization to prevent a recursed situation from being 
instantiated to begin with.

**Pull request merged**